devito/types/sparse.py

```diff
@@ -62,8 +66,24 @@ def __init_finalize__(self, *args, **kwargs):
     @classmethod
     def __indices_setup__(cls, *args, **kwargs):
         dimensions = as_tuple(kwargs.get('dimensions'))
+        # Need this not to break MatrixSparseFunction
+        try:
+            _sub_funcs = tuple(cls._sub_functions)
+        except TypeError:
+            _sub_funcs = ()
+        # If a subfunction provided use the sparse dimension
+        for f in _sub_funcs:
+            if f in kwargs:
+                try:
+                    sparse_dim = kwargs[f].indices[0]
+                    break
+                except AttributeError:
+                    continue
+        else:
+            sparse_dim = Dimension(name='p_%s' % kwargs["name"])
```

Comment:
This could be made the default value for `dimensions = as_tuple(kwargs.get('dimensions'))` which would shave off some lines by removing the need for the `if not dimensions` block. It would also allow the `if args` block to be moved to the top of the function, improving readability

Reply:
tweaked